c++/src/kj/compat/tls-test.c++

```diff
@@ -818,6 +822,241 @@ KJ_TEST("TLS client certificate verification") {
+
+  struct ClientRequest {
+    Maybe<Exception> maybeException;
+    Own<PromiseFulfiller<Own<AsyncIoStream>>> clientFulfiller;
+  };
+
+public:
+  MockConnectionReceiver(AsyncIoProvider& provider): provider(provider) {}
+
+  Promise<Own<AsyncIoStream>> accept() override {
+    return acceptImpl();
+  }
+
+  Promise<AuthenticatedStream> acceptAuthenticated() override {
+    return acceptImpl().then([](auto stream) -> AuthenticatedStream {
+      return { kj::mv(stream), LocalPeerIdentity::newInstance({}) };
+    });
+  }
+
+  uint getPort() override {
+    return 0;
+  }
+  void getsockopt(int, int, void*, uint*) override {}
+  void setsockopt(int, int, const void*, uint) override {}
+
+  Promise<Own<AsyncIoStream>> connect() {
+    // Mock a new successful connection to our receiver.
+    return connectImpl();
+  }
+
+  Promise<void> badConnect() {
+    // Mock a new failed connection to our receiver.
+    return connectImpl(KJ_EXCEPTION(DISCONNECTED, "Pipes are leaky")).ignoreResult();
+  }
+
+private:
+  Promise<Own<AsyncIoStream>> acceptImpl() {
+    if (clientRequests.empty()) {
+      KJ_ASSERT(!serverFulfiller);
+      auto paf = newPromiseAndFulfiller<void>();
+      serverFulfiller = kj::mv(paf.fulfiller);
+      return paf.promise.then([this] {
+        return acceptImpl();
+      });
+    }
+
+    // This is accepting in FILO order, it shouldn't matter in practice.
+    auto request = kj::mv(clientRequests.back());
+    clientRequests.removeLast();
+
+    KJ_IF_MAYBE(exception, kj::mv(request.maybeException)) {
+      request.clientFulfiller = nullptr;  // The other end had an issue, break the promise.
+      return kj::mv(*exception);
+    } else {
+      auto pipe = provider.newTwoWayPipe();
+      request.clientFulfiller->fulfill(kj::mv(pipe.ends[0]));
+      return kj::mv(pipe.ends[1]);
+    }
+  }
+
+  Promise<Own<AsyncIoStream>> connectImpl(Maybe<Exception> maybeException = nullptr) {
+    auto paf = newPromiseAndFulfiller<Own<AsyncIoStream>>();
+    clientRequests.add(ClientRequest{ kj::mv(maybeException), kj::mv(paf.fulfiller) });
+
+    if (auto fulfiller = kj::mv(serverFulfiller)) {
+      fulfiller->fulfill();
+    }
+
+    return kj::mv(paf.promise);
+  }
+
+  AsyncIoProvider& provider;
+
+  Own<PromiseFulfiller<void>> serverFulfiller;
+  Vector<ClientRequest> clientRequests;
+};
+
+class TlsReceiverTest final: public TlsTest {
+  // TlsReceiverTest augments TlsTest to test TlsConnectionReceiver.
+public:
+  TlsReceiverTest(): TlsTest() {
+    auto baseReceiverPtr = kj::heap<MockConnectionReceiver>(*io.provider);
+    baseReceiver = baseReceiverPtr.get();
+    receiver = tlsServer.wrapPort(kj::mv(baseReceiverPtr));
+  }
+
+  Own<ConnectionReceiver> receiver;
+  MockConnectionReceiver* baseReceiver;
+};
+
+KJ_TEST("TLS receiver basics") {
+  TlsReceiverTest test;
+  ErrorNexus e;
+
+  test.receiver->run();
+
+  auto clientPromise = test.baseReceiver->connect()
+                           .then([&](auto stream) {
```

Comment:
The indentation (and length of the next line) here is a bit extreme. As per [the style guide](https://github.com/capnproto/capnproto/blob/master/style-guide.md#spacing-and-bracing), we generally indent these kinds of `.then()` calls by either 0 or 4 spaces relative to the previous line. It does say "or line them up nicely with the previous line if it makes it easier to read", but I almost never see that done, especially not with this much indentation involved.

Reply:
Fair enough. I think I've convinced clang-format to play ball by penalizing whitespace more than assignment breaks. It adds an extra line...which probably couldn't hurt for comprehension anyway.

Reply:
Vitali did some work not too long ago on getting clang-format to work for our indentation preferences. I don't think it ever got finished, though, it looks like the internal PR for it is still open. Historically the extreme indentation of nested continuations is the main reason we haven't been able to use clang-format.